staking/contracts/Staking.sol

```diff
@@ -4,87 +4,250 @@ pragma solidity >=0.6.2 <0.8.0;
+
+    function __Staking_init(IERC20Upgradeable _token) external initializer {
         token = _token;
+        __Ownable_init_unchained();
     }
 
-    function createLock(address account, uint amount, uint slope, uint cliff) public {
-        //todo проверки
-        uint blockTime = roundTimestamp(block.timestamp);
-        BrokenLineDomain.Line memory line = BrokenLineDomain.Line(blockTime, getStake(amount, slope, cliff), slope);
-        BrokenLineDomain.Line memory lineLocked = BrokenLineDomain.Line(blockTime, amount, slope);
+    function stop() external onlyOwner notStopped {
+        stopped = true;
+    }
+
+    function startMigration(address to) external onlyOwner {
+        migrateTo = to;
+    }
+
+    function stake(address account, address delegate, uint amount, uint slope, uint cliff) external notStopped returns (uint) {
+        require(amount > 0, "amount negative");
+        require(cliff <= TWO_YEAR_WEEKS, "cliff too big");
+        require(amount.div(slope) <= TWO_YEAR_WEEKS, "period too big");
+        require(token.transferFrom(account, address(this), amount), "transfer failed");
 
-        totalSupplyLine.add(line, cliff);
-        locks[account].balance.add(line, cliff);
-        locks[account].locked.add(lineLocked, cliff);
+        counter++;
+
+        uint blockTime = roundTimestamp(block.timestamp);
+        (uint stAmount, uint stSlope) = getStake(amount, slope, cliff);
+        LibBrokenLine.Line memory line = LibBrokenLine.Line(blockTime, stAmount, stSlope);
+        totalSupplyLine.add(counter, line, cliff);
+        locks[delegate].balance.add(counter, line, cliff);
+        line = LibBrokenLine.Line(blockTime, amount, slope);
+        locks[account].locked.add(counter, line, cliff);
+        deposits[counter].locker = account;
+        deposits[counter].delegate = delegate;
         locks[account].amount = locks[account].amount.add(amount);
-        require(token.transferFrom(account, address(this), amount), "failure while transferring");
+        return counter;
+    }
 
-        // как меняется lock общий, когда юзер приходит/уходит/меняет
-        // 1. нужно применить пропущенные изменения (окончания локов)
-        // 2. если добавляем, то
+    function reStake(uint id, address newDelegate, uint newAmount, uint newSlope, uint newCliff) external notStopped returns (uint) {
+        address account = deposits[id].locker;
+        address delegate = deposits[id].delegate;
+        uint blockTime = roundTimestamp(block.timestamp);
+        verification(account, id, newAmount, newSlope, newCliff, blockTime);
+        removeLines(id, account, delegate, newAmount, blockTime);
+        return addLines(account, newDelegate, newAmount, newSlope, newCliff, blockTime);
     }
 
-    function totalSupply() public returns (uint) {
-        if (totalSupplyLine.initial.start == 0) { //no lock
+    function withdraw() external {
+        uint blockTime = roundTimestamp(block.timestamp);
```

Comment:
is it better to move this logic inside if (!stopped)?

Reply:
yes better to move logic in if()
agree. do it